src/decode/ykpt/packet_parser/mod.rs

```diff
@@ -0,0 +1,303 @@
+//! A packet parser for the Yk PT trace decoder.
+
+use crate::errors::HWTracerError;
+use deku::{bitvec::BitSlice, DekuRead};
+use std::{fmt::Write, iter::Iterator};
+
+mod packets;
+use packets::*;
+
+#[derive(Clone, Copy, Debug)]
+enum PacketParserState {
+    /// Initial state, waiting for a PSB packet.
+    Init,
+    /// The "normal" decoding state.
+    Normal,
+    /// We are decoding a PSB+ sequence.
+    PSBPlus,
+}
+
+impl PacketParserState {
+    /// Returns the kinds of packet that are valid for the state.
+    fn valid_packets(&self) -> &'static [PacketKind] {
```

Comment:
I'm not convinced by this function because in general I expect the follow-on packets to be somewhat conditional and then the `&'static` lifetimes won't work. I would expect that this will need to be inlined but, perhaps, PT is simple enough to get away with this?

Reply:
> I expect the follow-on packets to be somewhat conditional and then the &'static lifetimes won't work.

Can you elaborate. What's a follow-on packet? What's a conditional packet?

This function just says what *kinds* of packet are valid in each state, so that we don't waste effort decoding a packet which can never arise in the current parsing context. It isn't returning any packet data, so I do expect this to only ever return a static array.

Reply:
s/follow-on/valid/

I'd be moderately surprised if for a state `S` we can always statically calculate the valid packets with 100% accuracy (but maybe I'm wrong!).

Reply:
There's no calculation required. The Intel manual specifies when a given type of packet may come in.

Reply:
Just to show that I'm not making this up, here's an example:
![image](https://user-images.githubusercontent.com/604955/199066049-12fb38c7-102e-45e0-adbd-9afd032812bf.png)

Reply:
OK, let's worry about it later. I don't like the current design, but that's because I have written too many parsers, and this is a parser in disguise, but it's not too important now.